src/core/ucc_context.c

```diff
@@ -204,3 +205,84 @@ void ucc_context_config_print(const ucc_context_config_h config, FILE *stream,
             config->lib->full_prefix, (ucc_config_print_flags_t)flags);
     }
 }
+
+ucc_status_t ucc_context_create(ucc_lib_h lib,
+                                const ucc_context_params_t *params,
+                                const ucc_context_config_h  config,
+                                ucc_context_h *context)
+{
+    ucc_cl_context_params_t cl_ctx_params;
+    ucc_cl_context_t       *cl_ctx;
+    ucc_cl_lib_t           *cl_lib;
+    ucc_context_t          *ctx;
+    ucc_status_t            status;
+    uint64_t                i;
+    int                     num_cls;
+
+    num_cls = config->n_cl_cfg;
+    ctx     = ucc_malloc(sizeof(ucc_context_t), "ucc_context");
+    if (!ctx) {
+        ucc_error("failed to allocate %zd bytes for ucc_context",
+                  sizeof(ucc_context_t));
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    ctx->lib = lib;
+    memcpy(&ctx->params, params, sizeof(ucc_context_params_t));
+
+    ctx->cl_ctx = (ucc_cl_context_t **)ucc_malloc(
+        sizeof(ucc_cl_context_t *) * num_cls, "cl_ctx_array");
+    if (!ctx->cl_ctx) {
+        ucc_error("failed to allocate %zd bytes for cl_ctx array",
+                  sizeof(ucc_cl_context_t *) * num_cls);
+        status = UCC_ERR_NO_MEMORY;
+        goto error_ctx;
+    }
+    ctx->n_cl_ctx        = num_cls;
+    cl_ctx_params.params = params;
+
+    for (i = 0; i < num_cls; i++) {
+        cl_lib = config->configs[i]->cl_lib;
+        status = cl_lib->iface->context_create(cl_lib, &cl_ctx_params,
+                                               config->configs[i], &cl_ctx);
+        if (UCC_OK != status) {
+            ucc_error("failed to create cl context for %s",
+                      cl_lib->iface->super.name);
+            goto error_ctx_create;
+        }
+        ctx->cl_ctx[i] = cl_ctx;
+    }
+    ucc_debug("created ucc context %p", ctx);
```

Comment:
pls add lib name to debug print

Reply:
btw, should be info or debug? i'm not sure. Do we have any rule of thumb what is info/debug?

Reply:
feels like 'debug' to me - this doesn't really provide any information the user can benefit from... to me, info feels like non-trivial stuff, like the transports/components acutally selected during the run (something the user would have a hard time deducing).